funnel/models/membership.py

```diff
@@ -0,0 +1,371 @@
+# -*- coding: utf-8 -*-
+
+from sqlalchemy.ext.associationproxy import association_proxy
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.ext.hybrid import hybrid_property
+
+from coaster.sqlalchemy import immutable
+
+from . import BaseMixin, UuidMixin, db
+from .profile import Profile
+from .project import Project
+from .user import User
+
+__all__ = ['ProfileAdminMembership', 'ProjectCrewMembership']
+
+
+class ImmutableMembershipMixin(UuidMixin, BaseMixin):
+    """
+    Support class for immutable memberships
+    """
+
+    __uuid_primary_key__ = True
+    #: List of columns that will be copied into a new row when a membership is amended
+    __role_columns__ = ()
+    #: Parent column ('profile_id' or 'project_id' in the subclasses)
+    __parent_column__ = None
+
+    #: Start time of membership, ordinarily a mirror of created_at except
+    #: for records created when the member table was added to the database
+    granted_at = immutable(
+        db.Column(db.TIMESTAMP(timezone=True), nullable=False, default=db.func.utcnow())
+    )
+    #: End time of membership, ordinarily a mirror of updated_at
+    revoked_at = db.Column(db.TIMESTAMP(timezone=True), nullable=True)
+
+    @declared_attr
+    def revoked_by_id(cls):
+        """Id of user who revoked the membership"""
+        return db.Column(
+            None, db.ForeignKey('user.id', ondelete='SET NULL'), nullable=True
+        )
+
+    @declared_attr
+    def revoked_by(cls):
+        """User who revoked the membership"""
+        return db.relationship(User, foreign_keys=[cls.revoked_by_id])
```

Comment:
We need a matching `granted_by`, to record the identity of the person who created the membership.